Bpa configtypechanges

BennettAllen1 opened the pull request:
Hi,
please have a look at these changes I have made, but especially concentrating on those in the main where the class structure is called.
also for the etcd_plugins changes by adding the classes in the config_access folder
The unit tests relating to the validate function in config_access still need to be changed to take into affect the class as well as those for the main function where class_return is called.
Also the unit tests in test_config_type still needs work on checking the scripts for the shared_config validation

RobDover commented on src/metaswitch/clearwater/config_manager/test/test_config_access.py:
```diff
@@ -1031,6 +1045,41 @@ def test_missing_args(self):
+            return_value=[dns_json_config_plugin.DnsJson('path'),
+                          bgcf_json_config_plugin.BgcfJson('path'),
+                          shared_config_config_plugin.SharedConfig('path')])
+class TestClassReturn(unittest.TestCase):
```
You've written two tests in this class, but I don't think the second one is adding very much. Fundamentally it will run through the exact same underlying code logic each time.

This brings up an interesting question: how do we decide which tests are useful? The way I like to approach this is by asking: "What different types of inputs could this function receive?". Then once you've identified all the different types, you can write a test for each. To see how this works, let's go through this function as an example:

`class_return()` takes two parameters - `key` and `filepath`. 
- For the `key`, there are a bunch of possible inputs, but they will all be treated in one of two ways: either they will find the matching `ConfigType` class or they won't. Therefore, we should test both of these cases. 
- For the `filepath`, we expect that this will be passed into the `ConfigType` class somehow. We should test that it is passed through correctly.
- We should also think about whether there's any interaction between the inputs. In this case, I don't think there is.

The function also calls through to `load_plugins_from_dir()`. We should have think about what sort of inputs we could get from there, because `load_plugins_from_dir()` behaves differently depending on the state of the system. `class_return()` will behave the same way for most things it's possible for `load_plugins_from_dir()` to return. But there is one potentially interesting case - what happens when it returns nothing? Will this cause issues? We should probably check.

So from this, we get a list of four test cases we should run:
- valid key
- invalid key
- filepath set correctly on ConfigType object
- `load_plugins_from_dir()` returns no classes

Note that we can arrive at this list with limited understanding of what the actual code inside the function looks like. Furthermore, if you were to write these four tests, you would end up finding a bug in your code - it doesn't handle invalid keys correctly!

By taking this approach to unit tests, you can end up proactively finding bugs in your code. If you don't do it like this, you can sometimes find that writing the UTs ends up devolving into a box-ticking exercise.

RobDover commented on src/metaswitch/clearwater/config_manager/test/test_config_type.py:
```diff
@@ -0,0 +1,85 @@
+                  '$/usr/share/clearwater/clearwater-config-manager/scripts/config_validation/shared_ifcs_schema.xsd',
+                  '$path'], ]
+        list3 = []
+        list4 = ['script1', 'script2']
```
Rather than calling these `list1`, `list2`, etc., it would probably be better to call them `bgcf_scripts`, `ifcs_scripts`, etc.

RobDover commented on src/metaswitch/clearwater/config_manager/test/test_config_type.py:
```diff
@@ -0,0 +1,85 @@
+
+class TestConfigTypePluginLoader(unittest.TestCase):
+    # maybe tests already to test this for the function as used elsewhere
+    # as it is just the same
```
If we aren't using the same code as the other places, we should have our own test. The two might diverge.

RobDover commented on src/metaswitch/clearwater/config_manager/test/plugins/plugin_loader_test_plugin.py:
```diff
@@ -0,0 +1,40 @@
+_log = logging.getLogger("example_plugin")
+
+
+class PluginLoaderTestPlugin(SynchroniserPluginBase):
```
Somewhere in this file there should be a comment or a docstring that makes it clear that this class is only to be used for testing.

RobDover commented on src/metaswitch/clearwater/config_manager/config_type_class_plugin.py:
```diff
@@ -0,0 +1,57 @@
+
+class ConfigType:
+    """This is the base class for config type"""
+    def validate(self):
```
Having gone and had a look at the config plugins, there is no indication of what any of the fields that are set on them are for. Writing up that information here becomes doubly important.

RobDover commented on src/metaswitch/clearwater/config_manager/config_access.py:
```diff
@@ -343,19 +353,48 @@ def main(args):
+    # makes sense. Otherwise you would get always get refused by the arg
+    # parsing.
+    if config_classes == []:
+        raise NoConfigFound
```
Be careful when raising exceptions. You should only do it when one of the following is true:
- You have handling code that will catch the exception somewhere.
- You are comfortable with the exception being shown to whoever is running your script.

I don't think either of these two is true here at the moment. If you hit this exception, the user will be shown a Python stack trace - it probably won't look very nice!

But the question remains - how should we handle this circumstance? Have a look and see if you can figure out how we do this elsewhere in the script, then come and have a chat with me about it and we can go through it together.

RobDover commented on src/metaswitch/clearwater/config_manager/test/test_config_type.py:
```diff
@@ -0,0 +1,182 @@
+        self.assertListEqual(answer, ifcs_expected_script)
+
+    @mock.patch("metaswitch.clearwater.config_manager.config_type_class_plugin.os")
+    def test_get_shared(self, mock_os):
```
You aren't really testing that `get_sharedconfig_validation` function works the way you want it to here - all you're doing is checking that it looks in a directory at some point.

I would just scrap this and cover it as part of the `TestSharedValidation` tests below.

RobDover commented on src/metaswitch/clearwater/config_manager/config_access.py:
```diff
@@ -343,19 +353,48 @@ def main(args):
+    config_classes = load_plugins_in_dir(PLUGIN_DIR, config_location)
+    for choice in config_classes:
+        if config_type == choice.name:
+            return choice
```
Last time, I mentioned the [next](https://docs.python.org/2/library/functions.html#next) keyword. If you're not going to handle the case where none of the plugins have the right name, then it can greatly simplify your code here.

Have a look at what it is supposed to do and see if you can figure out how to use it. Come and talk to me if you get stuck.

RobDover commented on src/metaswitch/clearwater/config_manager/config_access.py:
```diff
@@ -343,19 +353,48 @@ def main(args):
+def lookup_config_type(config_type, config_location):
+    """
+    This takes the string defining the config_type from arg.parse and returns
+    a instance of the class containing data on the config type selected
```
A few issues to fix here:
- It's still unclear what `config_location` represents here. Why is it passed in?
- It feels a bit odd that you require that the config_type string comes from `arg.parse`. (a) I don't believe `arg.parse` is ever actually called anywhere in the code, and (b) ideally, your functions shouldn't care where their inputs come from.
- `a instance` -> `an instance`

RobDover commented on src/metaswitch/clearwater/config_manager/config_access.py:
```diff
@@ -343,19 +353,48 @@ def main(args):
+        if config_type == choice.name:
+            return choice
+        # The config_type will always be the choice.name of one class as
+        # parse_arguments will make sure config_type == choice.name
```
Comments should always be placed above the code they reference. Also, try not to use pseudo-code in comments where possible :).

RobDover commented on src/metaswitch/clearwater/config_manager/config_access.py:
```diff
@@ -343,19 +353,48 @@ def main(args):
+    The options for config type are designated by the plugins available
     :return:
     """
+    # Pick the options to appear in argument parsing, the config types are
```
`,` -> `.`

RobDover commented on src/metaswitch/clearwater/config_manager/config_access.py:
```diff
@@ -343,19 +353,48 @@ def main(args):
+    # If there are no config_classes loaded from the PLUGIN_DIR this means you
+    # will be unable to select anything, hence you want to raise an error, that
+    # makes sense. Otherwise you would get always get refused by the arg
+    # parsing.
```
Good pattern to try to catch yourself doing and fix: whenever you're writing a comment above an if statement and the comment includes a restatement of that if statement, it's almost always better to remove the restatement and move the comment inside the if block. Eg. rather than:
```
# If x is non-empty, do something
if x:
    do_something()
```
Instead write:
```
if x:
    # We have to do something because ...
    do_something()
```

RobDover commented on src/metaswitch/clearwater/config_manager/config_access.py:
```diff
@@ -422,85 +460,59 @@ def delete_outdated_config_files():
-    # in the above loop.
+    # in the above loop. The VALIDATION_SCRIPTS_FOLDER is in the ConfigType
+    # base class so see there, this is only ran through for shared_config
+    # validation
```
Rather than trying to fix this comment, I'd just get rid of it entirely. It's not normally good practice to leave comments about the past/future in code, because they're liable to become out of date very easily!

When someone does come to add something to prevent bash injection, they'll be able to look through our lovely, easy-to-read code and reach their own conclusions about where to make their changes. ;)

BennettAllen1 commented:
I have changed this to compare to master but the areas I have changed during bug fix (and since the last review) are as follows: 
config_access.py - selected_config as a instance gets passed into functions rather than the single string for name, in main and follows on from there.
parse_arguments - has had text formatting added but mainly the function config_error has been added to verify the config choice and offer the suggestion to look at another node.
the config now gets downloaded once and this string gets passed around to avoid being able to change the config file after validation. In ready_for_upload_checks and also has effect in write_config_to_etcd

config_type_class_plugin.py - validate function, especially the errors part

the Clearwater_etcd_plugins directory I have changed some text in help_text for some of them

RobDover commented on src/metaswitch/clearwater/config_manager/config_type_class_plugin.py:
```diff
@@ -0,0 +1,130 @@
+    def get_sharedconfig_validation(self):
+        """Returns the scripts to be used for shared config file validation."""
+        script_dir = glob.glob(os.path.join(VALIDATION_SCRIPTS_FOLDER, '*'))
+        # We can only execute scripts that have execute permissions.:q
```
Rogue `:q`

RobDover commented on src/metaswitch/clearwater/config_manager/config_type_class_plugin.py:
```diff
@@ -0,0 +1,130 @@
+        for script in self.scripts:
+            try:
+                log.debug("Running validation script %s", script)
+                subprocess.check_output(self.scripts[script], stderr=subprocess.STDOUT)
```
When you are looping over a dictionary and have to do something like this, it's often cleaner to re-write the loop like this:
```
for name, script in self.scripts.items():
    function_that_uses_script(script)
    function_that_uses_name(name)
```
The `items()` method turns the dictionary into a list of tuples, making the loop look a bit nicer.

It's not a big deal though - at this point it's not worth fixing.
